src/java.base/share/classes/java/nio/Buffer.java

```diff
@@ -823,7 +824,7 @@ public MemorySegment bufferSegment(Buffer buffer) {
                 }
 
                 @Override
-                public Runnable acquireSession(Buffer buffer, boolean async) {
+                public Runnable acquireSessionOrNull(Buffer buffer, boolean async) {
```

Comment:
If allocation/performance is an issue, a relatively straightforward way to adjust the code would be to let go of the TWR entirely. E.g. we have code that does this:

```
Buffer b = ...
try {
    // use buffer.address();
} finally {
    Reference.reachabilityFence(b);
}
```

We could transform to this:

```
Buffer b = ...
acquire(b); // calls MemorySessionImpl.acquire0 if session is not null
try {
    // use buffer.address();
} finally {
    release(b); // calls MemorySessionImpl.release0 if session is not null (and maybe adds a reachability fence, just in case)
}
```
This leads to a simpler patch that is probably easier to validate. The remaining IOUtil code will be using a different idiom, and perhaps we can, at some point, go back and make that consistent too.

Reply:
The AutoCloseable experiment was interesting to try but I think you are right, acquire try { .. } finally release would be simpler and also remove the concerns about the performance critical code paths.